pdns/iputils.cc

```diff
@@ -543,3 +547,56 @@ void setSocketSendBuffer(int fd, uint32_t size)
 {
   setSocketBuffer(fd, SO_SNDBUF, size);
 }
+
+std::set<std::string> getListOfNetworkInterfaces()
+{
+  std::set<std::string> result;
+#if HAVE_GETIFADDRS
+  struct ifaddrs *ifaddr;
+  if (getifaddrs(&ifaddr) == -1) {
+    return result;
+  }
+
+  for (struct ifaddrs *ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
+    if (ifa->ifa_name == nullptr) {
+      continue;
+    }
+    result.insert(ifa->ifa_name);
+  }
+
+  freeifaddrs(ifaddr);
+#endif
+  return result;
+}
+
+std::vector<ComboAddress> getListOfAddressesOfNetworkInterface(const std::string& itf)
+{
+  std::vector<ComboAddress> result;
+#if HAVE_GETIFADDRS
+  struct ifaddrs *ifaddr;
+  if (getifaddrs(&ifaddr) == -1) {
+    return result;
+  }
+
+  for (struct ifaddrs *ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
+    if (ifa->ifa_name == nullptr || strcmp(ifa->ifa_name, itf.c_str()) != 0) {
+      continue;
```

Comment:
Couldn't find a better spot for this comment. If the user passes in a name that does not exist, the output is an empty table. Should it be an error? (This is really a question, I'm not suggesting either way is better):

```
> getListOfAddressesOfNetworkInterface('foo')
{}
```

Reply:
Good question! I kind of like the current behaviour, but that's likely because I only care about interfaces that do have addresses assigned to them and I see how someone could have a different opinion. On the other hand raising an exception might be a bit too much, especially since on some setups interfaces come and go.

Reply:
That makes sense, let's leave it.